### DISCUSSION TITLE
GPU Memory changes by TP/PP and recompute-activations when the GPU-NUM is stable

### DISCUSSION THREAD

**Discussion by Listen-WLS** [NVIDIA/Megatron-LM#1486] [Q&A]
As mentioned in the title, when the number of Gpus remains the same, I change the size of TP/PP and get the following GPU memory usage. After I read the article "https://arxiv.org/pdf/2205.05198.pdf", I think the GPU memory footprint mainly includes the following several parts:
(1) Parameter occupation
(2) Optimizer occupation
(3) Gradient occupation
(4) Activation value occupation

I used 8 Gpus to train the Llama3-8B model. When batchsize=1, the GPU memory occupancy can be expressed as:
1) Parameter occupancy:  8B* 2/(TP* PP)=2GB  
2) the optimizer takes up:  8B* 4 /（TP* PP） +  8 B*  4 /（TP* PP）+  8 B*  4 /（TP* PP）= 12 GB 
3) Gradient occupancy:   8B/PP* 2=2GB
4) Activation occupation:  Batchsize * sequenc_length  *  hiddensize *  2 = 2.15 GB

I have done many sets of experiments to prove the corresponding situation between this GPU memory formula and the truth GPU memory occupation, but the real GPU memory occupation will gradually increase with the change of PP. I would like to know which of the above four GPU memory occupation is affected by the growth of PP.

![5030a1165b0987fffe3093a5f2702a6e_78145](https://github.com/user-attachments/assets/8d252e1a-07a0-4770-9f0b-77003a5ea6c0)
![dcdd6f9a07b1f17cdd8d0d4255481604_51808](https://github.com/user-attachments/assets/ab22a17f-58f8-4896-b8b0-4b3d794fb5ac)

**Comment by Advait251206** [comment]
Your back-of-the-envelope memory model is missing several components that become more visible as you change the TP/PP configuration while keeping the total GPU count fixed.

The biggest point is that **activation memory is not simply**

\[
\text{BS} \times \text{SeqLen} \times \text{Hidden} \times 2
\]

especially in Megatron-LM with pipeline parallelism and activation checkpointing.

For an 8B Llama model, the memory on each GPU is more like

\[
M =
M_{\text{params}}
+
M_{\text{optimizer}}
+
M_{\text{grads}}
+
M_{\text{activations}}
+
M_{\text{communication}}
+
M_{\text{workspace}}
+
M_{\text{allocator fragmentation}}
\]

where the last three terms are often several GB.

## 1. Parameters and optimizer

Your estimates are approximately correct.

For TP = \(t\), PP = \(p\):

- Parameters scale as **1 / (TP × PP)**.
- Adam states scale as **1 / (TP × PP)**.
- FP32 master weights (used with mixed precision) also scale as **1 / (TP × PP)**.

These components are **not** the reason for the increasing memory because `TP × PP = 8` in all of your experiments.

---

## 2. Gradients

Gradients are slightly more subtle than the simplified formula.

Depending on the training stage, Megatron may simultaneously hold:

- Gradient tensors
- Gradient accumulation buffers
- NCCL reduction buffers

Therefore, the gradient footprint is not always exactly

\[
8B \times 2 / PP
\]

There are transient buffers during backward and communication that are not included in the analytical formula.

---

## 3. Activation memory

This is the biggest source of the discrepancy.

With

```bash
--recompute-granularity full
--recompute-method block
--recompute-num-layers 32
```

Megatron does **not** recompute everything.

It still stores several tensors for backward, including:

- Layer inputs
- Residual connections
- LayerNorm inputs
- FlashAttention metadata
- Various tensors required for recomputation
- Pipeline communication activations

Therefore activation memory is much larger than

\[
\text{Batch} \times \text{SeqLen} \times \text{Hidden} \times 2
\]

and its size depends on the model partitioning.

---

## 4. Pipeline parallelism introduces additional buffers

When `PP > 1`, each pipeline stage maintains extra memory for:

- Input activation buffers
- Output activation buffers
- Send buffers
- Receive buffers
- Pipeline scheduling queues

During the 1F1B schedule, multiple microbatches can remain resident simultaneously.

Even with

```bash
--micro-batch-size 1
```

these pipeline buffers still exist.

Therefore, pipeline parallelism introduces memory that is **not captured** by the simple activation formula.

---

## 5. Tensor parallelism changes GEMM workspaces

As TP decreases, each GPU performs larger matrix multiplications.

Libraries such as:

- cuBLASLt
- FlashAttention

allocate temporary workspaces whose size depends on GEMM dimensions.

These workspaces are **not** part of:

- Parameters
- Optimizer states
- Gradients
- Activations

yet they can easily consume several hundred MB (or even more depending on sequence length).

---

## 6. CUDA caching allocator fragmentation

`nvidia-smi` reports **reserved GPU memory**, not only actively used memory.

PyTorch's CUDA caching allocator keeps previously allocated memory reserved for reuse.

Different TP/PP configurations produce different allocation patterns, so

```python
torch.cuda.memory_allocated()
```

may remain similar while

```python
torch.cuda.memory_reserved()
```

changes by several GB because of allocator fragmentation.

This effect becomes more noticeable with long sequence lengths such as 8192.

---

## 7. Communication buffers

Megatron allocates persistent communication buffers for:

- Tensor Parallel all-reduce
- Reduce-scatter
- Pipeline send/receive
- Sequence Parallel (if enabled)

These buffers also vary with TP/PP and are not included in the simplified analytical model.

---

# Why memory increases from TP=1, PP=8 to TP=8, PP=1

Your measurements are

| TP | PP | GPU Memory (GB) |
|----|----|----------------:|
| 1 | 8 | 19.06 |
| 2 | 4 | 21.41 |
| 4 | 2 | 22.39 |
| 8 | 1 | 23.78 |

Although parameter memory remains essentially constant because

\[
TP \times PP = 8,
\]

the following components increase:

- Larger activation footprint per GPU because each rank owns more transformer layers.
- Larger temporary GEMM and FlashAttention workspaces.
- Larger temporary backward buffers.
- Different CUDA allocator reservation and fragmentation behavior.

In particular, when `PP = 1`, every GPU owns the entire transformer stack. Even with activation recomputation enabled, many more intermediate tensors exist within the same process than when the model is split across eight pipeline stages.

Activation checkpointing **reduces** activation memory, but it does **not** make activation memory completely independent of the number of layers assigned to each GPU.

Therefore, the increase from roughly **19 GB → 24 GB** is expected.

---

# How to verify this

To understand where the extra memory is coming from, compare memory snapshots between TP/PP configurations using

```python
torch.cuda.memory_summary()
```

or

```python
torch.cuda.memory_snapshot()
```

If you're using PyTorch 2.4 or newer, you can also use

```python
torch.cuda.memory._snapshot()
```

which provides a detailed allocation graph.

Megatron's `report_memory()` utility is also useful for determining whether the additional memory comes from activations, communication buffers, or allocator reservations.

---

# Conclusion

The additional GPU memory is **not** coming from a single one of the four categories in your simplified model.

Parameter memory, optimizer states, and gradients largely follow the expected scaling. The discrepancy is primarily due to:

- Activation-related tensors that remain even with checkpointing
- Pipeline communication buffers
- Temporary GEMM and FlashAttention workspaces
- NCCL communication buffers
- CUDA caching allocator fragmentation

These components depend on the TP/PP configuration and become increasingly important for long sequence lengths such as **8192**, which explains why the measured memory differs from the simplified analytical estimate.

---

If this answer helped or pointed you in the right direction, I'd appreciate it if you could mark it as the accepted answer so it's easier for others with the same issue to find.

Also, if you found my contribution useful, I'd appreciate it if you could check out my GitHub profile, follow me, and star any repositories you find interesting.

GitHub: https://github.com/Advait251206
